src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/Macros/CoalesceMacro.cs

```diff
@@ -50,27 +52,33 @@ public IMacroConfig CreateConfig(IEngineEnvironmentSettings environmentSettings,
 
         public void EvaluateConfig(IEngineEnvironmentSettings environmentSettings, IVariableCollection vars, IMacroConfig config)
         {
-            CoalesceMacroConfig realConfig = config as CoalesceMacroConfig;
+            CoalesceMacroConfig? realConfig = config as CoalesceMacroConfig;
 
             if (realConfig == null)
             {
                 throw new InvalidCastException("Unable to cast config as a CoalesceMacroConfig");
             }
 
-            object targetValue = null;
-            if (vars.TryGetValue(realConfig.SourceVariableName, out object currentSourceValue) && !Equals(currentSourceValue ?? string.Empty, realConfig.DefaultValue ?? string.Empty))
+            object? targetValue = null;
+            if (!string.IsNullOrEmpty(realConfig.SourceVariableName)
+                && vars.TryGetValue(realConfig.SourceVariableName!, out object currentSourceValue)
+                && !Equals(currentSourceValue ?? string.Empty, realConfig.DefaultValue ?? string.Empty))
             {
                 targetValue = currentSourceValue;
             }
             else
             {
-                if (!vars.TryGetValue(realConfig.FallbackVariableName, out targetValue))
+                if (!string.IsNullOrEmpty(realConfig.FallbackVariableName)
+                    && !vars.TryGetValue(realConfig.FallbackVariableName!, out targetValue))
                 {
                     environmentSettings.Host.Logger.LogDebug("Unable to find a variable to fall back to called " + realConfig.FallbackVariableName);
                     targetValue = realConfig.DefaultValue;
                 }
             }
-            vars[config.VariableName] = targetValue?.ToString();
+            if (targetValue is not null)
+            {
+                vars[config.VariableName] = targetValue.ToString();
```

Comment:
I'm thinking if we should add a null check into the `VariableCollection` (untill all the callers use nullables, so that we can rely on build time check that values in `VariableCollection` are not being assigned nulls)